### ISSUE TITLE
Opcode flags

### ISSUE THREAD

**Issue by stoklund** [stoklund/cretonne#12]
Cretonne instruction opcodes should have _flags_ that provide useful information for the verifier and optimizer. The flags would indicate constant properties of an opcode, such as:
- `is_terminator` - This is a terminator instruction where control can never pass through, such as an unconditional branch, return, or trap instruction. Every EBB must end with a terminator, and terminators can't appear inside an EBB.
- `is_branch` - This is a branch instruction that can transfer control to another EBB.
- `can_trap` - This instruction can trap, causing control flow to diverge. Besides the obvious `trap` instructions, division by zero and memory access instructions can also trap. This does not make the instruction a terminator since the trapping can be conditional.

In the Python meta instruction descriptions, flags are passed as extra keyword arguments to the `Instruction` constructor:

``` python
brz = Instruction(
        'brz', r"""
        Branch when zero.

        If ``c`` is a :type:`b1` value, take the branch when ``c`` is false. If
        ``c`` is an integer value, take the branch when ``c = 0``.
        """,
        ins=(c, EBB, args), is_branch=True)
```

The `gen_instr.py` script should be modified to emit additional opcode methods, one for each flag:

``` rust
impl Opcode {
    pub fn is_branch(self) -> bool {
        match self {
            Opcode::Brz => true
            ...
            _ => false
        }
    }
}
```

These flags should be generated from instruction definitions so that a target ISA can define custom terminator instructions. For example, the x86-64 ISA would define a `ud2` instruction that is marked as a terminator because it always traps.

**Comment by stoklund** [comment]
@dinfuehr, do you think you could eliminate the old `is_terminating()` method before I close this issue?

**Comment by dinfuehr** [comment]
Yes, but I have a question for that:

Is there already a way to get the `Opcode` from the `InstructionData` (except for writing a big `match` myself)? If not, I would add a function `fn opcode(&self) -> Opcode` to `InstructionData` that returns the Opcode with the help of pattern matching.

**Comment by stoklund** [comment]
There is, but that method is generated by the Python scripts. You can see it in the generated docs

**Comment by stoklund** [comment]
Fixed completely by #33.
